Automattic/jetpack
# Boost: Try: Add an e2e test base framework

#### Changes proposed in this Pull Request:
This PR is my very first take on implementing an e2e testing framework for the Boost plugin. It is heavily inspired and has a lot of logic lifted from the jetpack e2e test framework structure.

This attempt might not be 100% correct as it might not fit in a wider scheme to  introduce e2e testings for projects in the monorepo but it can start the discussions.

There are a few things to take in consideration with the POC:

- As mentioned above this attempt is lifting a lot of logic from the Jetpack e2e setup. However I have taken only what is (for the moment) relevant for the Jetpack Boost plugin as Jetpack e2e setup as a lot of complexity (around connections, modules, search, etc...) that is not applicable to Jetpack Boost. In lifting the code, I realised that there could be some core logic that could be shared across multiple plugins for implementing the e2e framework for each (if relevant) plugins in the monorepo but it might be difficult to find a way to share code easily.
- Because I did not have access to decrypt the e2e config local-test,js file (from `projects/plugins/jetpack/tests/e2e/config/encrypted.enc`), i created and committed a `test.js` config file for Boost which does not contain any credentials. As a result, I had to create a plugin that bypass the jetpack connection (which will limit our test capabilities since Boost relies on an active Jetpack connection to get Speed Score via the WPCOM API). So I guess we would have to introduce the same logic used by Jetpack e2e framework at some point.
- The whole 2e2 testing is using the LocalTunnel capability since ultimately we need to have WPCOM communicate with the Jetpack Boost e2e test instance. However because wp-env environment is only using http (for example http://localhost:8889) but by default LocalTunnel is using https, the UI of the e2e test environment is failing (because non HTTPs resources cannot be failing) so I resulted to use the **non HTTPs** version of the LocalTunnel url. Meaning all the test are done under http and not https but there is no confidential data transmitted. I am not sure yet to know how to resolve this.
- The current e2e tests created are very simple and not too relevant yet because I did not want to spend too much time doing some meaningful test yet (and to learn more about playwright) as this might end up not being used if it is not the appropriate approach.

All in all, I think, with more work and cleanup, this approach could work but it is not really re-usable across plugins.

I would love to have your feedback and guidance @brbrr, @adimoldovan, @jeherve  on this first attempt.

@thingalon, @eAdnan007, I added you also as reviewer so you have visibility on the discussions and progress.

#### Jetpack product discussion
N/A

#### Does this pull request change what data or activity we track or use?
N/A

#### Testing instructions:
From the `projects/plugins/boost` folder run:
* `pnpm test-e2e:start`: this will start the e2e testing environment.
* `pnpm test-e2e:run`: this will run the (very basic) e2e tests. Check they are all passing.
* `pnpm test-e2e:stop`: this will stop the e2e testing environment.

Check that the whole process is working.
